### PR TITLE
XRDEV-1605 - encrypted backups

### PR DESCRIPTION
**Pull request by amoog**
Encrypting/decrypting backups using gpg and automatically generated key.

### CODE REVIEW

File: src/packages/src/xroad/ubuntu/generic/xroad-proxy.postinst [hunk 1]
```diff
@@ -77,6 +77,14 @@ case "$1" in
 
   db_stop
 
+  # generate gpg keypair when needed - directory does not exist and server is initialised
+  if [ ! -d /etc/xroaqd/gpghome ] ; then
```
Comment (jhyoty) [inline src/packages/src/xroad/ubuntu/generic/xroad-proxy.postinst:81]: typo in the directory name

File: src/packages/src/xroad/redhat/SPECS/xroad-proxy.spec [hunk 2]
```diff
@@ -184,6 +185,16 @@ if [ $1 -gt 1 ]; then
     rm -rf %{_localstatedir}/lib/rpm-state/%{name}
 fi
 
+if [ $1 -gt 1 ]; then
+  # upgrade, generate gpg keypair when needed
+  if [ ! -d /etc/xroaqd/gpghome ] ; then
```
Comment (jhyoty) [inline src/packages/src/xroad/redhat/SPECS/xroad-proxy.spec:190]: typo in path?

File: src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/backup_xroad_proxy_configuration.sh [hunk 3]
```diff
@@ -20,9 +20,21 @@ OPTIONS:
...
+get_proxy_prop() {
+  # local.ini overrides keys
+  local value=$(crudini --get /etc/xroad/conf.d/local.ini "$2" "$3" 2>/dev/null || echo EMPTYKEY)
+  if [ -n "$value" ]; then
+    if [ $value = EMPTYKEY ]; then
+      local value=$(crudini --get /etc/xroad/conf.d/"$1" "$2" "$3" 2>/dev/null || echo -n "$4")
+    fi
+  fi
+  echo $value
+}
```
Comment (jhyoty) [inline src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/backup_xroad_proxy_configuration.sh:36]: Alternatively, avoids treating EMPTYKEY a special value
```
get_proxy_prop() {
  local value
  if ! value=$(crudini --get /etc/xroad/conf.d/local.ini "$2" "$3" 2>/dev/null); then
    value=$(crudini --get /etc/xroad/conf.d/"$1" "$2" "$3" 2>/dev/null || echo "$4")
  fi
  echo "$value"
}
```

**Review by jhyoty — CHANGES_REQUESTED** [review]

File: src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/backup_xroad_proxy_configuration.sh (hunk 3 above)
Comment (amoog) [reply src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/backup_xroad_proxy_configuration.sh:36]: Yes this is definitely better, my bash skills are really lacking.

File: src/packages/src/xroad/common/base/usr/share/xroad/scripts/_backup_restore_common.sh [hunk 4]
```diff
@@ -117,4 +117,15 @@ has_command () {
     command -v $1 &>/dev/null
 }
 
+get_proxy_prop() {
+  # local.ini overrides keys
+  local value=$(crudini --get ~/etc/conf.d/local.ini "$2" "$3" 2>/dev/null || echo EMPTYKEY)
+  if [ -n "$value" ]; then
+    if [ $value = EMPTYKEY ]; then
```
[the remark this answers is not in the document]
Comment (amoog) [reply src/packages/src/xroad/common/base/usr/share/xroad/scripts/_backup_restore_common.sh:124]: Changed this in many places in common and proxy folders.

File: src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/get_security_server_id.sh [hunk 5]
```diff
@@ -6,4 +6,10 @@ HOST=$(crudini --get /etc/xroad/db.properties '' serverconf.hibernate.connection
 PORT=$(crudini --get /etc/xroad/db.properties '' serverconf.hibernate.connection.url | cut -d '/' -f 3 | cut -d ':' -f2)
 
 export PGPASSWORD=${PW}
-psql -t -A -F / -h ${HOST:-localhost} -p ${PORT:-5432} -d serverconf -U ${USER:-serverconf} -c "select identifier.xroadinstance, identifier.memberclass, identifier.membercode, serverconf.servercode from serverconf inner join client on serverconf.owner=client.id inner join identifier on client.identifier=identifier.id;"
+psql -t -A -F / -h ${HOST:-localhost} -p ${PORT:-5432} -d serverconf -U ${USER:-serverconf} -c \
```
[the remark this answers is not in the document]
Comment (amoog) [reply src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/get_security_server_id.sh:9]: True. But what to do? This crude parsing seems very fragile, I see two options 1) write server id to some file during initialisation 2) add some service to proxy that can be called from localhost without authentication. Both seem to be outside the scope of this ticket.

[the remark this answers is not in the document]
Comment (jhyoty) [reply src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/get_security_server_id.sh:9]: There is more complete database property parsing in https://github.com/nordic-institute/X-Road/blob/develop/src/packages/src/xroad/common/proxy/usr/share/xroad/scripts/restore_db.sh

File: src/packages/src/xroad/redhat/SPECS/xroad-proxy.spec [hunk 6]
```diff
@@ -184,6 +185,16 @@ if [ $1 -gt 1 ]; then
     rm -rf %{_localstatedir}/lib/rpm-state/%{name}
 fi
 
+if [ $1 -gt 1 ]; then
+  # upgrade, generate gpg keypair when needed
+  if [ ! -d /etc/xroaq/gpghome ] ; then
```
Comment (jhyoty) [inline src/packages/src/xroad/redhat/SPECS/xroad-proxy.spec:190]: Should probably be `/etc/xroad/gpghome`?

File: src/packages/src/xroad/ubuntu/generic/xroad-proxy.postinst [hunk 7]
```diff
@@ -77,6 +77,14 @@ case "$1" in
 
   db_stop
 
+  # generate gpg keypair when needed - directory does not exist and server is initialised
+  if [ ! -d /etc/xroaq/gpghome ] ; then
```
Comment (jhyoty) [inline src/packages/src/xroad/ubuntu/generic/xroad-proxy.postinst:81]: Should probably be `/etc/xroad/gpghome`

**Review by jhyoty — COMMENTED** [review]
In addition, _uploading backups via the gui or api_ now fails since the api rejects gpg files as invalid tar files.

**Review by jhyoty — CHANGES_REQUESTED** [review]
There are build failures in proxy-ui-api module (checkstyle does not pass, some test failures).

**Review by jhyoty — CHANGES_REQUESTED** [review]

**Review by jhyoty — APPROVED** [review]

**Review by jhyoty — APPROVED** [review]

**Pull request merged**